plugins/obs-filters/nvidia-audiofx-filter.c

```diff
@@ -0,0 +1,924 @@
+	ng->frames = frames;
+	ng->channels = channels;
+
+	/* allocate buffers */
+	ng->copy_buffers[0] = bmalloc(frames * channels * sizeof(float));
+	ng->nvafx_segment_buffers[0] =
+		bmalloc(NVAFX_FRAME_SIZE * channels * sizeof(float));
+	for (size_t c = 1; c < channels; ++c) {
+		ng->copy_buffers[c] = ng->copy_buffers[c - 1] + frames;
+		ng->nvafx_segment_buffers[c] =
+			ng->nvafx_segment_buffers[c - 1] + NVAFX_FRAME_SIZE;
+	}
+
+	/* reserve cicular buffers */
+	for (size_t i = 0; i < channels; i++) {
+		circlebuf_reserve(&ng->input_buffers[i],
+				  frames * sizeof(float));
+		circlebuf_reserve(&ng->output_buffers[i],
+				  frames * sizeof(float));
+	}
+
+	/* create resampler if the source is not at 48 kHz */
+	if (sample_rate == NVAFX_SAMPLE_RATE) {
+		ng->nvafx_resampler = NULL;
+		ng->nvafx_resampler_back = NULL;
+	} else {
+		struct resample_info src, dst;
+		src.samples_per_sec = sample_rate;
+		src.format = AUDIO_FORMAT_FLOAT_PLANAR;
+		src.speakers = nv_convert_speaker_layout((uint8_t)channels);
+
+		dst.samples_per_sec = NVAFX_SAMPLE_RATE;
+		dst.format = AUDIO_FORMAT_FLOAT_PLANAR;
+		dst.speakers = nv_convert_speaker_layout((uint8_t)channels);
+
+		ng->nvafx_resampler = audio_resampler_create(&dst, &src);
+		ng->nvafx_resampler_back = audio_resampler_create(&src, &dst);
+	}
+
+	nvidia_audio_update(ng, settings);
+	return ng;
+}
+
+static inline void process_fx(struct nvidia_audio_data *ng)
+{
+
+	/* Resample if necessary */
+	if (ng->nvafx_resampler) {
+		float *output[MAX_PREPROC_CHANNELS];
+		uint32_t out_frames;
+		uint64_t ts_offset;
+		audio_resampler_resample(ng->nvafx_resampler,
+					 (uint8_t **)output, &out_frames,
+					 &ts_offset,
+					 (const uint8_t **)ng->copy_buffers,
+					 (uint32_t)ng->frames);
+
+		for (size_t i = 0; i < ng->channels; i++) {
+			for (ssize_t j = 0,
+				     k = (ssize_t)out_frames - NVAFX_FRAME_SIZE;
+			     j < NVAFX_FRAME_SIZE; ++j, ++k) {
+				if (k >= 0) {
+					ng->nvafx_segment_buffers[i][j] =
+						output[i][k];
+				} else {
+					ng->nvafx_segment_buffers[i][j] = 0;
+				}
+			}
+		}
+	} else {
+		for (size_t i = 0; i < ng->channels; i++) {
+			for (size_t j = 0; j < NVAFX_FRAME_SIZE; ++j) {
+				ng->nvafx_segment_buffers[i][j] =
+					ng->copy_buffers[i][j];
+			}
+		}
+	}
+
+	/* Execute */
+	size_t runs = ng->has_mono_src ? 1 : ng->channels;
+	if (ng->reinit_done) {
+		pthread_mutex_lock(&ng->nvafx_mutex);
+		for (size_t i = 0; i < runs; i++) {
+			NvAFX_Status err = NvAFX_Run(
+				ng->handle[i], &ng->nvafx_segment_buffers[i],
+				&ng->nvafx_segment_buffers[i],
+				ng->num_samples_per_frame, ng->num_channels);
+			if (err != NVAFX_STATUS_SUCCESS) {
+				if (err == NVAFX_STATUS_FAILED) {
+					do_log(LOG_DEBUG,
+					       "NvAFX_Run() failed, error NVAFX_STATUS_FAILED.\n"
+					       "This can occur when changing the FX and is not consequential.");
+					os_atomic_set_bool(
+						&ng->reinit_done,
+						false); // stop all processing; this will be reset at new init
```

Comment:
Just put the comment before its associated line.

Reply:
done